### PR TITLE
Don’t fear the reap peer

### PR DESCRIPTION
**Pull request by kriskowal**

### CODE REVIEW

File: service-proxy.js [hunk 1]
```diff
@@ -444,6 +456,14 @@ function refreshServicePeer(serviceName, hostPort) {
 
     // The old way: fully connect every egress to all affine peers.
     peer.connectTo();
+
+    // Unmark recently seen peers, so they don't get reaped
+    delete self.peersToReap[hostPort];
+    // Mark known peers, so they are candidates for future reaping
+    if (self.connectedPeers[hostPort]) {
```
Comment (jcorbin) [inline service-proxy.js]: if `!`!

Comment (Raynos) [inline service-proxy.js]: if (!...)

File: service-proxy.js [hunk 2]
```diff
@@ -37,6 +37,7 @@ var SERVICE_PURGE_PERIOD = 5 * 60 * 1000;
 var DEFAULT_MIN_PEERS_PER_WORKER = 5;
 var DEFAULT_MIN_PEERS_PER_RELAY = 5;
 var DEFAULT_STATS_PERIOD = 30 * 1000; // every 30 seconds
+var DEFAULT_REAP_PEERS_PERIOD = 60 * 1000; // every 5 minutes
```
Comment (Raynos) [inline service-proxy.js]: 5 * 60 * 1000

File: service-proxy.js [hunk 3]
```diff
@@ -748,6 +768,51 @@ function requestPeriodicStats() {
...
+    var self = this;
+
+    self.logger.info('Reaping dead peers', self.extendLogInfo({
+        peers: self.peersToReap
+    }));
+
+    var peersToReap = Object.keys(self.peersToReap);
+    for (var i = 0; i < peersToReap.length; i++) {
+        var hostPort = peersToReap[i];
+        var peer = self.channel.peers.get(peer);
```
Comment (jcorbin) [inline service-proxy.js]: arg should be `hostPort`

Comment (jcorbin) [inline service-proxy.js]: check if peer maybe?

File: service-proxy.js [hunk 4]
```diff
@@ -748,6 +768,51 @@ function requestPeriodicStats() {
...
+        for (var j = 0; j < peer.connections.length; j++) {
+            var connection = peer.connections[j];
+            if (connection.direction === 'out') {
```
Comment (jcorbin) [inline service-proxy.js]: why filter, why not just always drain

File: service-proxy.js [hunk 5]
```diff
@@ -748,6 +768,51 @@ function requestPeriodicStats() {
...
+
+        var serviceNames = Object.keys(self.peersToReap[hostPort]);
+        for (var j = 0; j < serviceNames.length; j++) {
+            var serviceName = serviceNames[j];
+            var service = self.getServiceChannel(serviceName);
+            if (!service) {
+                continue;
+            }
+            service.peers.delete(hostPort);
+        }
```
Comment (Raynos) [inline service-proxy.js]: Please also remove peer from top channel :)

File: service-proxy.js [hunk 6]
```diff
@@ -748,6 +768,51 @@ function requestPeriodicStats() {
...
+
+    self.logger.info('Reaping dead peers', self.extendLogInfo({
+        peers: self.peersToReap
+    }));
+
+    var peersToReap = Object.keys(self.peersToReap);
+    for (var i = 0; i < peersToReap.length; i++) {
+        var hostPort = peersToReap[i];
+        var peer = self.channel.peers.get(peer);
+        for (var j = 0; j < peer.connections.length; j++) {
```
Comment (jcorbin) [inline service-proxy.js]: I'd be really happy if we could factor out a `.reapPeer(peer)`

File: service-proxy.js [hunk 7]
```diff
@@ -748,6 +768,51 @@ function requestPeriodicStats() {
...
+            var connection = peer.connections[j];
+            if (connection.direction === 'out') {
+                connection.drain('reaped for expired advertisement');
+            }
+        }
+
+        var serviceNames = Object.keys(self.peersToReap[hostPort]);
+        for (var j = 0; j < serviceNames.length; j++) {
+            var serviceName = serviceNames[j];
+            var service = self.getServiceChannel(serviceName);
```
Comment (jcorbin) [inline service-proxy.js]: call this var consistent with other passages

File: service-proxy.js (hunk 5 above)
Comment (anson627) [inline service-proxy.js]: potential refactoring: code can be shared with removeServicePeer.

**Comment by kriskowal** [comment]
Look, ma, test. @jakev @jcorbin

File: service-proxy.js [hunk 8]
```diff
@@ -95,10 +96,20 @@ function ServiceDispatchHandler(options) {
     self.periodicStatsTimer = null;
     self.statsPeriod = options.statsPeriod || DEFAULT_STATS_PERIOD;
 
+    self.peersToReap = Object.create(null);
+    self.connectedPeers = Object.create(null);
+    self.reapPeersTimer = null;
+    self.reapPeersPeriod = options.reapPeersPeriod || DEFAULT_REAP_PEERS_PERIOD;
```
Comment (jcorbin) [inline service-proxy.js]: let's hook this up to a remote config thingy

Comment (jakev) [inline service-proxy.js]: Why am I on this? :)

Comment (kriskowal) [inline service-proxy.js]: You’re not @raynos? Hi. :facepalm:

Comment (jakev) [inline service-proxy.js]: Nope, but it looks like we have quite similar names! I'll see myself out, carry on!;)

**Comment by kriskowal** [comment]
Verbal :ok: from @Raynos and @jcorbin pending a remote config flag for release to production. PR forthcoming.

**Pull request merged**